### PR TITLE
Auto refant

### PR DESCRIPTION
**Pull request by Athanaseus**
# Auto select reference antenna

### Criteria:
- The antenna should have a maximum distance to the centre of the array that is less than `maxdist`.
- The antenna should have baselines that are greater than `minbase`.
- Once above are satisfied, get top 3 (or 2 or 1) antennas with the least amount of flags.

NB: refant can take a list of antennas. e.g. refant='m000,m001,m007'
(This will use antenna m000, and m001 and m007 as alternates if/when m000 drops out)

### CODE REVIEW

**Comment by paoloserra** [comment]
@Athanaseus this is not working for me. I get the following error:

```
2020-09-22 07:43:11 CARACal.Stimela.antenna_flag_summary INFO: job started at 2020-09-22 07:43:11.541834
# 6e8e2339ea702d84e4c5f58081ff62cbb4afbfa9ceac6fb4d83d4319efc61d95
# Traceback (most recent call last):
#   File "/stimela_mount/code/run.py", line 27, in <module>
#     value = list(map(int, value))
# TypeError: 'bool' object is not iterable
2020-09-22 07:43:13 CARACal.Stimela.antenna_flag_summary ERROR: docker returns error code 1
2020-09-22 07:43:13 CARACal.Stimela.antenna_flag_summary ERROR: job failed at 2020-09-22 07:43:13.410296 after 0:00:01.868462
2020-09-22 07:43:13 CARACal ERROR: Job 'antenna_flag_summary:: Flagging summary  ms=gps1-cal.ms' failed: docker returns error code 1 [PipelineException]
2020-09-22 07:43:13 CARACal INFO:   More information can be found in the logfile at output/logs-20200922-074258/log-caracal.txt
2020-09-22 07:43:13 CARACal INFO: exiting with error code 1
```

and, in the logfile:

```
2020-09-22 07:43:11 CARACal.Stimela.antenna_flag_summary INFO: Starting container [antenna_flag_summary-14033143981796816007605796939259]. Timeout set to -1. The container ID is printed below.
# running docker start -a antenna_flag_summary-14033143981796816007605796939259
# Traceback (most recent call last):
#   File "/stimela_mount/code/run.py", line 27, in <module>
#     value = list(map(int, value))
# TypeError: 'bool' object is not iterable
2020-09-22 07:43:13 CARACal.Stimela.antenna_flag_summary ERROR: docker returns error code 1
2020-09-22 07:43:13 CARACal.Stimela.antenna_flag_summary ERROR: job failed at 2020-09-22 07:43:13.410296 after 0:00:01.868462
2020-09-22 07:43:13 CARACal.Stimela.antenna_flag_summary ERROR: Traceback (most recent call last):
2020-09-22 07:43:13 CARACal.Stimela.antenna_flag_summary ERROR:   File "/home/pserra/caracal-venv/lib/python3.6/site-packages/stimela/recipe.py", line 704, in run
2020-09-22 07:43:13 CARACal.Stimela.antenna_flag_summary ERROR:     job.run_job()
2020-09-22 07:43:13 CARACal.Stimela.antenna_flag_summary ERROR:   File "/home/pserra/caracal-venv/lib/python3.6/site-packages/stimela/recipe.py", line 424, in run_job
2020-09-22 07:43:13 CARACal.Stimela.antenna_flag_summary ERROR:     self.job.start(output_wrangler=self.apply_output_wranglers)
2020-09-22 07:43:13 CARACal.Stimela.antenna_flag_summary ERROR:   File "/home/pserra/caracal-venv/lib/python3.6/site-packages/stimela/docker.py", line 171, in start
2020-09-22 07:43:13 CARACal.Stimela.antenna_flag_summary ERROR:     kill_callback=lambda: utils.xrun("docker", ["kill", self.name]))
2020-09-22 07:43:13 CARACal.Stimela.antenna_flag_summary ERROR:   File "/home/pserra/caracal-venv/lib/python3.6/site-packages/stimela/utils/xrun_poll.py", line 227, in xrun
2020-09-22 07:43:13 CARACal.Stimela.antenna_flag_summary ERROR:     raise StimelaCabRuntimeError("{} returns error code {}".format(command_name, status))
2020-09-22 07:43:13 CARACal.Stimela.antenna_flag_summary ERROR: stimela.utils.StimelaCabRuntimeError: docker returns error code 1
2020-09-22 07:43:13 CARACal.Stimela.crosscal INFO: Completed jobs : ['version-ps_crosscal_before-ms0', 'set_model_cal-0']
2020-09-22 07:43:13 CARACal.Stimela.crosscal INFO: Remaining jobs : []
2020-09-22 07:43:13 CARACal.Stimela.crosscal INFO: Saving pipeline information in .last_crosscal.json
2020-09-22 07:43:13 CARACal ERROR: Job 'antenna_flag_summary:: Flagging summary  ms=gps1-cal.ms' failed: docker returns error code 1 [PipelineException]
2020-09-22 07:43:13 CARACal INFO:   More information can be found in the logfile at output/logs-20200922-074258/log-caracal.txt
2020-09-22 07:43:13 CARACal INFO:   You are running version 1.1.0
2020-09-22 07:43:13 CARACal ERROR: Traceback (most recent call last):
2020-09-22 07:43:13 CARACal ERROR:   File "/home/pserra/caracal-venv/lib/python3.6/site-packages/caracal/main.py", line 184, in __run
2020-09-22 07:43:13 CARACal ERROR:     pipeline.run_workers()
2020-09-22 07:43:13 CARACal ERROR:   File "/home/pserra/caracal-venv/lib/python3.6/site-packages/caracal/workers/worker_administrator.py", line 397, in run_workers
2020-09-22 07:43:13 CARACal ERROR:     worker.worker(self, recipe, config)
2020-09-22 07:43:13 CARACal ERROR:   File "/home/pserra/caracal-venv/lib/python3.6/site-packages/caracal/workers/crosscal_worker.py", line 574, in worker
2020-09-22 07:43:13 CARACal ERROR:     prefix_msbase, label=label, ftype="primary")
2020-09-22 07:43:13 CARACal ERROR:   File "/home/pserra/caracal-venv/lib/python3.6/site-packages/caracal/workers/crosscal_worker.py", line 326, in solve
2020-09-22 07:43:13 CARACal ERROR:     do_KGBF(i)
2020-09-22 07:43:13 CARACal ERROR:   File "/home/pserra/caracal-venv/lib/python3.6/site-packages/caracal/workers/crosscal_worker.py", line 136, in do_KGBF
2020-09-22 07:43:13 CARACal ERROR:     pipeline.maxdist[iobs], i)
2020-09-22 07:43:13 CARACal ERROR:   File "/home/pserra/caracal-venv/lib/python3.6/site-packages/caracal/workers/utils/manage_antennas.py", line 17, in get_refant
2020-09-22 07:43:13 CARACal ERROR:     recipe.run()
2020-09-22 07:43:13 CARACal ERROR:   File "/home/pserra/caracal-venv/lib/python3.6/site-packages/stimela/recipe.py", line 755, in run
2020-09-22 07:43:13 CARACal ERROR:     raise PipelineException(exc, self.completed, job, self.remaining) from None
2020-09-22 07:43:13 CARACal ERROR: stimela.exceptions.PipelineException: Job 'antenna_flag_summary:: Flagging summary  ms=gps1-cal.ms' failed: docker returns error code 1
2020-09-22 07:43:13 CARACal INFO: exiting with error code 1
```

I've tried to debug but got nowhere in the time I could spend on it, sorry.

**Comment by Athanaseus** [comment]
Sorry for the confusion @paoloserra this PR depends on https://github.com/ratt-ru/Stimela/pull/678

**Comment by gigjozsa** [comment]
Test this please.

**Comment by paoloserra** [comment]
> Sorry for the confusion @paoloserra this PR depends on [ratt-ru/Stimela#678](https://github.com/ratt-ru/Stimela/pull/678)

OK @Athanaseus , with that Stimela branch it runs through.

I'm puzzled by the results though. If I set `minbase: 200` the first choice for a reference antenna is m008, which forms baselines shorter than 200 m with m007, m010, m040. Did I misunderstand something or is there a bug somewhere?

**Comment by paoloserra** [comment]
@Athanaseus of course, the reason for my issue with m008  above is that my  MS did not have all MeerKAT antennas, and m008  was a perfectly fine choice indeed. My bad, sorry!

I'm now checking against a different MS and will probably approve soon.

**Comment by paoloserra** [comment]
OK, just ran a few tests, it's good. I will approve but we shouldn't merge until your Stimela branch is merged and included into a Stimela release. So I'm adding a "do not merge" label.

**Review by paoloserra — APPROVED** [review]

**Comment by Athanaseus** [comment]
> I'm puzzled by the results though. If I set minbase: 200 the first choice for a reference antenna is m008, which forms baselines shorter than 200 m with m007, m010, m040. Did I misunderstand something or is there a bug somewhere?

Ah yes you a right @paoloserra, thanks, this was definitely a bug. At the worst-case scenario if the criteria set doesn't return any antenna this should log and crush that ref ant should be specified manually. Or otherwise just set `maxdist` to the furthest antenna from the centre (or a very high number that includes most antenna) and this will at least still check for flags percentages.

I have made the appropriate changes to the branch, please retest.

**Comment by Athanaseus** [comment]
Also, should this be default in the config files?

File: caracal/workers/crosscal_worker.py
Comment (paoloserra) [inline caracal/workers/crosscal_worker.py:140]: This line  should cause CARACal to exit.

Currently it does not, the pipeline continues with  `refant=''` and crashes.

**Review by paoloserra — CHANGES_REQUESTED** [review]

File: caracal/workers/crosscal_worker.py
Comment (paoloserra) [reply caracal/workers/crosscal_worker.py:140]: I meant L.140 of course.

Comment (paoloserra) [reply caracal/workers/crosscal_worker.py:140]: And yes, I agree that `refant: auto` should  be the default.

**Comment by o-smirnov** [comment]
retest this please

**Comment by o-smirnov** [comment]
Stimela release is done, so if y'all want to proceed with a review and merge once the tests pass...

**Review by paoloserra — APPROVED** [review]

**Comment by paoloserra** [comment]
retest this please

**Comment by o-smirnov** [comment]
So with the default settings, ``refant=auto`` broke one of my old reductions. I can see it was using ``refant="m010,m022,m023"``. But the bpcal phases came out all sad:

image

Reverting to my sentimental favourite ``m008``, everything is fine:

image

So, I dunno.

**Comment by paoloserra** [comment]
Have you tried to manually set to m010?

**Comment by o-smirnov** [comment]
Aha, an excellent idea, thanks. Watch this space.

**Comment by o-smirnov** [comment]
OK, I'm coming to realize that the wideband 1GC is extremely sensitive to initial flag state. The best starting point for Caracal seems to be CAM flags only. In this case, I had an MS that was mildly flagged (~30% level overall) from a previous Caracal run... from before we got the Glorious Sacred Readonly Workflow going... and all sorts of crazy shit was going on at the 1GC stage. 

Finally recreated it from the archive, and voila, the crazy problems go away, including the one described above. Moral of the story is, watch your flags, kids. More is not always better.

One slight annoyance: even on virginal data, m008 does slightly better than the auto-selected refant (m010,m022,m023 as above). In the latter case I virtually lose one antenna (m047):

|m008|auto|
|-|-|
|image|image|

(The other two I had pre-flagged by hand due to wonky bandpasses). Just rotten luck?

I really detest the whole concept of a reference antenna. Reference antennas are the Trumps of radio astronomy. Pick them manually or pick them automatically, they're still orange and psychotic. #RefantsMustFall. Rescue us @JSKenyon, when you're done with your talk, we **really** need KGB CubiCal...

**Pull request merged**
